### PR TITLE
Sl 306

### PR DESCRIPTION
**Pull request by dmdesimone**

### CODE REVIEW

File: configuration/pih/liquibase/liquibase-sierraleone.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <changeSet id="20220411-fix-encounters-not-slotted-in-visits" author="ddesimone">
+         <preConditions onFail="MARK_RAN">
+             <sqlCheck expectedResult="0">
```
Comment (mogoodrich) [inline configuration/pih/liquibase/liquibase-sierraleone.xml:10]: I always get this backwards myself, but isn't this backwards?  I believe this precondition says, "if the query below returns 0, then just skip this change set  and mark it as having run"

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	from encounter e
+	inner join encounter_type et on et.encounter_type_id = e.encounter_type 
+	left outer join obs o on o.encounter_id = e.encounter_id 
+	where e.visit_id is null
+		and e.voided = 0
+		and encounter_type not in
+		(select encounter_type_id from encounter_type where uuid in
+		('10db3139-07c0-4766-b4e5-a41b01363145','1545d7ff-60f1-485e-9c95-5740b8e6634b','873f968a-73a8-4f9c-ac78-9f4778b751b6',
+		'4d77916a-0620-11e5-a6c0-1697f925ec7b','5b1b4a4e-0084-4137-87db-dba76c784439','74cef0a6-2801-11e6-b67b-9e71128cae77','d5ca53a7-d3b5-44ac-9aa2-1491d2a4b4e9',
+		'873f968a-73a8-4f9c-ac78-9f4778b751b6','39C09928-0CAB-4DBA-8E48-39C631FA4286','b3a0e3ad-b80c-4f3f-9626-ace1ced7e2dd'))
```
Comment (mogoodrich) [inline configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:14]: I didn't double-check all these encounter types... :)

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	void_reason = 'empty encounter with no visit';
+
+delete from temp_no_visits where obs_count = 0;
+
+-- gather all visits same day as those encounters
+drop temporary table if exists temp_visits;
+CREATE temporary table temp_visits
+	select v.visit_id , v.patient_id , v.date_started , v.date_stopped
+	from visit v
+	inner join temp_no_visits t on t.patient_id = v.patient_id  and date(v.date_stopped) = date(t.encounter_datetime) ;
```
Comment (mogoodrich) [inline configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:34]: The assumption here is that we don't need/care about multiple-day visits, as we are matching on date_stopped?  iie, what if the visit when from Apr 10 to April 12 and there was an encounter on April 11th?  Shouldn't it be slotted in that visit?

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+inner join temp_visits v on v.visit_id =
+	(select v2.visit_id from temp_visits2 v2
+	where v2.patient_id = t.patient_id
+	and date(v2.date_stopped) = date(t.encounter_datetime)
+	and v2.date_stopped < t.encounter_datetime 
+	order by v2.date_stopped desc limit 1
+	)
+set t.visit_id = v.visit_id,
+	t.date_started = v.date_started,
+	t.date_stopped = v.date_stopped
```
Comment (mogoodrich) [inline configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:66]: isn't this the case where we should be extended the visit date stopped out to the encounter date?

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+where t.visit_id is null;
+
+create index temp_fixed_visits_ei on temp_fixed_visits(encounter_id);
+
+-- expand visit date_stopped
+update visit v 
+inner join temp_fixed_visits t on t.visit_id = v.visit_id
+set v.date_stopped = t.encounter_datetime,
+	date_changed = now(),
+	changed_by = (select user_id from users where system_id = 'admin');
```
Comment (mogoodrich) [inline configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:76]: Okay, actually, this handles the question I had above... but, now, won't this potentially be inadvertently shrinking the date stopped in some cases?  Don't we only want to do this in the cases where v.date_stopped < t.encounter_datetime?

**Review by mogoodrich — COMMENTED** [review]
@dmdesimone  I had some questions, see below, but I'm probably just not reading the query right...

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql (hunk 3 above)
Comment (dmdesimone) [reply configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:34]: Good point - I made the assumption that visits at Wellbody were all same day, but I checked and it is NOT the case, so I modified the logic here

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql (hunk 4 above)
Comment (dmdesimone) [reply configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:66]: absolutely, oops!  Fixed this.

File: configuration/pih/liquibase/sql/fix_encounters_without_visits.sql (hunk 5 above)
Comment (dmdesimone) [reply configuration/pih/liquibase/sql/fix_encounters_without_visits.sql:76]: absolutely, oops! Fixed this.

File: configuration/pih/liquibase/liquibase-sierraleone.xml (hunk 1 above)
Comment (dmdesimone) [reply configuration/pih/liquibase/liquibase-sierraleone.xml]: this is the main thing I needed the review on... I thought so also but I modeled it after one already committed.  But anyway, this should fix it.

Comment (mogoodrich) [reply configuration/pih/liquibase/liquibase-sierraleone.xml]: it's *possible* I had a it wrong... :)

File: configuration/pih/liquibase/liquibase-sierraleone.xml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <changeSet id="20220411-fix-encounters-not-slotted-in-visits" author="ddesimone">
+         <preConditions onFail="MARK_RAN">
+             <not>
+	     	<sqlCheck expectedResult="0">
```
Comment (mogoodrich) [inline configuration/pih/liquibase/liquibase-sierraleone.xml:11]: did this change?  I double-checked, and I still think that this should actually check if the count is greater than zero?

**Pull request merged**